syntax = "proto3";

package auth;

// Данные пользователя, необходимые для открытия новой сессии.
message Credentials {
    string user_name = 1;
    string password = 2;
}

// Публичная информация о сессии, которая передается между сервисами.
message Cookie {
    int64 session_id = 1;
    string hash_key = 2;
}

message Status {
    enum Code {
        UNKNOWN = 0;
        OK = 1;
        ACCESS_DENIED = 2;
        INVALID_SESSION = 3;
    }
    Code code = 1;
}

message LoginRequest {
    Credentials credentials = 1;
}
message LoginResponse {
    Status status = 1;
    Cookie cookie = 2;
}

message LogoutRequest {
    Cookie cookie = 1;
}
message LogoutResponse {
}

message ValidationRequest {
    Cookie cookie = 1;
    string resource = 2;
}
message ValidationResponse {
    Status status = 1;
}

service Authenticator {
    rpc Login (LoginRequest) returns (LoginResponse) {}
    rpc Logout (LogoutRequest) returns (LogoutResponse) {}
    rpc Validate (ValidationRequest) returns (ValidationResponse) {}
}
